azavea/nasa-hyperspectral
# AVIRIS L2 Activator

# Overview

This PR implements a Python activator that takes an AVIRIS STAC
Item Id and performs the following steps in order:

1. Pull refl archive from FTP
2. Unzip archive in temp dir
3. Convert all HDR files in archive to DEFLATE COGs
4. Create a STAC Item
5. Add STAC Asset for each created COG
6. Upload each COG to S3
7. Create STAC Item by POST to Franklin
8. Print new STAC Item URL

Script exits 0 on success, non-zero on any failure by throwing exceptions. We can adjust later once it's clearer what the specifics of the environment we'll be running in look like, but for now we have an adaptable baseline.

Closes #55 

## Demo

In this first screenshot, we trigger the job but get an error:

![Screen Shot 2020-12-07 at 6 21 37 PM](https://user-images.githubusercontent.com/1818302/101490364-3c6e3100-3930-11eb-9ee9-29f92002ba43.png)

With the options we've selected (`--temp-dir data --keep-temp-dir`) we can resume where we left off after fixing the bug:

![Screen Shot 2020-12-07 at 6 05 14 PM](https://user-images.githubusercontent.com/1818302/101490361-3c6e3100-3930-11eb-8bbf-09f189a4930c.png)

I also provided `--skip-large` which skips translate of the 8GB HDR file, which would take ~40mins on my laptop.

## Notes

The dev instructions below provide some options to the task that allow incremental retries so that a developer doesn't have to continually redownload or specify the location of the tar.gz file. Don't forget to clean up whatever temp dir you set when done developing.

There are some improvements to be made to the developer workflow for these, but I hesitated to do so now, before knowing what the infrastructure will look like. Some of these tasks include:
- [x] Add Franklin instance to docker-compose and configure for running pipelines
- [ ] ~~Adjust error handling for chosen pipeline infrastructure~~
- [x] Add a STRTA for running a pipeline job

Running the activator job in MacOS Docker takes _forever_ and Docker spins the CPU the whole time, presumably due to work being done to the volume mount when I set `--temp-dir ./data`. If you don't care about intermediate results, you can run the job without this arg and it should go faster, but if you want to re-run the job you'll need to download the FTP archive each time.

Running a job locally is an exercise in patience. We'll very quickly need to migrate running this job to Batch even in dev, or further improve the local experience.

## Testing Instructions

Follow the new steps in the README to setup the dev env. Copied here that's:
```
./scripts/setup
./scripts/server
./scripts/catalogs-aviris
```

Then you should be able to hit http://localhost:9090 and see the AVIRIS collection.

Finally run the task:
`./scripts/run-activator-aviris-l2 aviris_f130329t01p00r06_sc01 --temp-dir ./data --skip-large --keep-temp-dir`
...and go bake a cake.

When the activator finally finishes, you should be able to visit the URL printed by the script and see a STAC Item contained in a new `aviris-l2-cogs` collection!

## Checklist
